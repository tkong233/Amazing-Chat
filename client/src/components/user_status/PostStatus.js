import React from "react";
import "./ViewStatus.css";
import { Button, Box } from "@material-ui/core";
import { connect } from "react-redux";
import { doPostStatus } from "../../actions/authActions";

class PostStatus extends React.Component {
  constructor() {
    super();
    this.state = {
      image: "",
      imageURL: "",
      text: "",
    };
  }

  onImageChange = (e) => {
    e.preventDefault();
    this.setState({
      imageURL: URL.createObjectURL(e.target.files[0]),
      image: e.target.files[0]
    });
  };

  onTextChange = (e) => {
    e.preventDefault();
    this.setState ({
        text: e.target.value,
    });
  }

  onSubmit = (e) => {    
    if (!this.state.image && !this.state.text) {
        // TODO: add alert message
        console.log("no image or text");
        return;
    }
    if (!this.state.image || !this.state.text) {
      alert("You have to add both text and image!");
      return;
    }
    const formData = new FormData();
    formData.append("NonTextfile", this.state.image);
    formData.append("text", this.state.text);
    const { email } = this.props.user;
    this.props.doPostStatus(formData, email, this.props.history);
  };

  render() {
    return (
      <div className="container" data-test = "postStatusComponent" style={{marginTop: "50px"}}>
        <div className="row">
          <form className="col s12">
            <div className="row">
              <div className="input-field col s12">
                <input id="last_name" type="text" className="validate" onChange={this.onTextChange}/>
                <label htmlFor="last_name">Your status</label>
              </div>
            </div>
            <div className="row">
              <Box display="flex" justifyContent="space-around">
                <Button
                  style={{ color: "#455a64", backgroundColor: "white" }}
                  variant="contained"
                  component="label"
                >
                  Choose an image
                  <input
                    type="file"
                    name="file"
                    style={{ display: "none" }}
                    onChange={this.onImageChange}
                  />
                </Button>
              </Box>
            </div>
            <div className="row">
              <Box display="flex" justifyContent="space-around">
                {this.state.imageURL ? (
                  <img
                    style={{
                      maxWidth: "50%",
                      maxHeight: "50%",
                      borderRadius: "8px",
                    }}
                    src={this.state.imageURL}
                    alt=""
                  />
                ) : (
                  ""
                )}
                {/* {this.state.text} */}
              </Box>
            </div>
            <div className="row">
              <Box display="flex" justifyContent="space-around" p={2}>
                <Button id="button"
                  style={{ color: "white", backgroundColor: "#455a64" }}
                  variant="contained"
                  onClick={this.onSubmit}
                >
                  Post!
                </Button>{" "}
              </Box>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
    user: state.auth.user,
  });

export default connect(mapStateToProps, { doPostStatus })(PostStatus);

