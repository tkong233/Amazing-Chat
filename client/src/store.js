import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { persistStore, persistReducer } from 'redux-persist'
import rootReducer from "./reducers";
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web

const initialState = {};

const persistConfig = {
  key: 'root',
  storage,
}

const middleware = [thunk];

const persistedReducer = persistReducer(persistConfig, rootReducer, initialState);

const store = createStore(

  persistedReducer,
  compose(
    applyMiddleware(...middleware),
    (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__()) ||
      compose
  )
);

const persistor = persistStore(store);

export {
  store,
  persistor,
  middleware
};



