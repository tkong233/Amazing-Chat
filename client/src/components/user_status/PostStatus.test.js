import React from 'react';
import PostStatus from './PostStatus';
import {shallow} from 'enzyme';
import checkPropTypes from 'check-prop-types';
import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16'; 
import { applyMiddleware, createStore } from 'redux';
import {middleware} from '../../store';
import rootReducer from '../../reducers/index'

Enzyme.configure({
    adapter: new EnzymeAdapter(),
});

const setUp = (initialState={}) =>{

    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    const store = createStoreWithMiddleware(rootReducer, initialState);
    const wrapper = shallow(<PostStatus store={store} />).dive();
    // console.log(wrapper.debug());
    return wrapper;
}

describe('PostStatus Component', ()=>{
    describe('Checking PropTypes', ()=>{
        it('Should not throw a warning', ()=>{
            const expectedProps = {
                user: {},
            }
            // eslint-disable-next-line react/forbid-foreign-prop-types
            const propsErr = checkPropTypes(PostStatus.propTypes, expectedProps, 'props', PostStatus.name);
            expect(propsErr).toBeUndefined();
        })
    });
    describe('Renders', ()=>{
        let wrapper;
        beforeEach(()=>{
            const initialState = {
                auth:{user: {
                    name: 'testuser',
                    email: 'testuser@test.com',
                }}
            };
            wrapper = setUp(initialState);
        });

        it('Should render without errors', ()=>{
            const container = wrapper.find(`[data-test='postStatusComponent']`);
            expect(container.length).toBe(1);
        });

        it('Snapshot testing', ()=>{
            expect(wrapper).toMatchSnapshot();
        });

        it('upload image', ()=>{
            global.URL.createObjectURL = jest.fn();
            wrapper.find('input').at(1).simulate('change', {
                preventDefault: () => {},
                target: {
                    files:['imageFile']
                }
            });
            expect(wrapper.state('image')).toBe('imageFile');
        });

        it('writing status', ()=>{
            const value = 'testStatusMessage'
            wrapper.find('input').at(0).simulate('change', {
                preventDefault: () => {},
                target: {
                    value
                }
            });
            expect(wrapper.state('text')).toBe(value);
        });

        it('fail to post -- incomplete status message or image', ()=>{
            wrapper.find('#button').simulate('click', {
                preventDefault: () => {},
            });
            expect(wrapper.state('text')).toStrictEqual('');
        })

    });
});