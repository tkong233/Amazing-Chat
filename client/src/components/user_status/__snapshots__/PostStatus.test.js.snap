// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostStatus Component Renders Snapshot testing 1`] = `ShallowWrapper {}`;
